assert: add strict functionality export

BridgeAR opened the pull request:
To further improve assert this adds a strict export. When used, all functions will use the strict equality instead of a loose one.

At the same time this is a doc only deprecation of the legacy loose mode. It is not intended to remove the old functionality as it is wildly used a lot and removing it does not seem to be a option.
I also improved the assert documentation in general a bit. It now outlines the used rules a bit better and clearly separates loose and strict equality.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
doc,assert

Trott commented on doc/api/assert.md:
```diff
@@ -6,6 +6,44 @@
 
 The `assert` module provides a simple set of assertion tests that can be used to
 test invariants.
+A `strict` and a `legacy` mode exist, while it is recommended to only use
```
Micro-nit: Add a blank line before this one?

Trott commented on doc/api/assert.md:
```diff
@@ -6,6 +6,44 @@
+
+## Strict mode
+
+When using the strict mode any assert function will use the equality used in the
```
Micro-nit: comma after `mode`?

Trott replied:
Micro-nit: backticks around `strict mode` and `assert`?

Trott commented on doc/api/assert.md:
```diff
@@ -6,6 +6,44 @@
+## Strict mode
+
+When using the strict mode any assert function will use the equality used in the
+strict function mode. So [`assert.deepEqual`][] will for example work the same
```
Micro-nit: comma before and after `for example`?

Trott commented on doc/api/assert.md:
```diff
@@ -6,6 +6,44 @@
+
+> Stability: 0 - Deprecated: Use strict mode instead.
+
+When accessing `assert` directly instead of using the `strict` property the
```
Nit: comma after `property`?

Trott commented on doc/api/assert.md:
```diff
@@ -6,6 +6,44 @@
+
+When accessing `assert` directly instead of using the `strict` property the
+"non-strict" equality will be used for the "non-strict" function names e.g.
+[`assert.deepEqual`][].
```
Nit: Put everything after `e.g.` and up to the closing `.` in parentheses.

Nit: Here and several places above, function names should be followed by `()` I believe. So `assert.deepEqual()` rather than `assert.deepEqual`. That said, if there's already a lot of no-parentheses examples in the doc, never mind. :-D

Trott commented on doc/api/assert.md:
```diff
@@ -6,6 +6,44 @@
+
+It is recommended to use the [ strict mode`][] instead as the loose equality
+checks are often not sufficient.
+
```
Nit: They're called `non-strict`, `legacy`, and `loose` throughout the document. I'd suggest picking just one of those and sticking with it.

Nit: `are often not sufficient` might be better expressed as `can often have surprising results`.

BridgeAR replied:
I rephrased it and hope it is clearer now :-)

BridgeAR commented:
@Trott I addressed all comments. I know it might seem a strong move but we do not have to remove the legacy mode ever and that is also why I only wanted a doc only deprecation.
Having the `strict mode` will likely improve the situation for many people though.

vsemozhetbyt commented on doc/api/assert.md:
````diff
@@ -7,6 +7,54 @@
+const assert = require('assert');
+```
+
+It is recommended to use the [ strict mode`][] instead as the
````
``[ strict mode`][]``: space instead of a backtick.

joyeecheung commented:
Ping @BridgeAR this needs a rebase.

BridgeAR commented:
Rebased

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/11605/

jasnell commented:
Needs another rebase :-)

BridgeAR commented:
Rebased plus some tiny documentation improvements while doing so (it is actually the SameValue comparison used by Object.is() and not the Object.is() comparison).

addaleax commented:
Landed in 06ab6f2f13981babe71f3a9fe577bebc3c7cb5ac, a319e90807bbc74b6d0e85ee9bec697acb68ebcb

MylesBorins commented:
This does not land cleanly in `v9.x`. Please feel free to manually backport by following [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md). Please also feel free to replace do-not-land if it is being backported

BridgeAR commented:
Backport opened in #19230

MylesBorins commented:
Requested backport to 8.x in #19230
